### ISSUE TITLE
Wishlist: detect if --8086 should be used 

### ISSUE THREAD

**Issue by hackerb9** [upx/upx#530]
## What's the problem (or question)?
I expected UPX to automatically detect that that .EXE I was compressing was compiled for 8086 processors. It did not and the resulting executable did not work on my machine (IBM PC 5155).

## What should have happened?
UPX is already magic for most every architecture and format. Is there no way to get any inkling whether a .EXE is compiled for 16-bit 8086? 

## Do you have an idea for a solution?

If what I'm asking is impossible, it would be nice if there was at least a heuristic that gave a warning to users when they use UPX on a file that seems _likely_ to be 8086 code.

## How can we reproduce the issue?
This is a known and well documented issue in UPX.

## Please tell us details about your environment.

* UPX version used (`upx --version`):
* Host Operating System and version: Debian GNU/Linux 
* Host CPU architecture: AMD64
* Target Operating System and version: MS-DOS 3
* Target CPU architecture: 8086

**Comment by jreiser** [comment]
Please state the failure message or symptoms.  Please exhibit an actual .exe for 8086 that fails.  Upload it ("Attach files by ..."), give a URL, or email it.  Compress with gzip or zip to reduce space and time on an upload.
As a work-around, dId you try the `--8086` command-line argument to UPX?

**Comment by hackerb9** [comment]
I apologize for not making it clear. This is a feature request, not a bug report. UPX is working exactly as described in the manual: if you do not give it the --8086 option, then UPX will render an 8086 executable no longer able to run on an 8086. Currently, --8086 is required. I'm hoping that someday UPX will be able to automatically detect the format. 

Here's the 8086 executable I am compressing: [kermit.exe](https://github.com/hackerb9/mrpiggy/releases/download/v0.0.2/kermit.exe).

**Comment by jreiser** [comment]
This feature request cannot be implemented.  An "8086-only .exe" has no marking which says so, and even inspecting the instructions to be compressed can prove only that at least 80286 is required, but not that 80286 instructions are forbidden.

Also, the difference between using `--8086` and not using it, is only 11 (eleven) bytes for kermit.exe, and only slightly more for .sys or .com files.  In the UPX `src` directory:
```
$ grep 8086 *.h *.cpp
conf.h:#define UPX_F_ELKS_8086         17              // NOT IMPLEMENTED
options.h:        CPU_8086 = 1,
p_com.h:    //virtual const char *getFullName(const options_t *o) const override { return o && o->cpu == o->CPU_8086 ? "i086-dos16.com" : "i286-dos16.com"; }
p_exe.h:    //virtual const char *getFullName(const options_t *o) const { return o && o->cpu == o->CPU_8086 ? "i086-dos16.exe" : "i286-dos16.exe"; }
p_sys.h:    //virtual const char *getFullName(const options_t *o) const override { return o && o->cpu == o->CPU_8086 ? "i086-dos16.sys" : "i286-dos16.sys"; }
help.cpp:                    "  --8086              make compressed com work on any 8086\n"
help.cpp:                    "  --8086              make compressed exe work on any 8086\n"
help.cpp:                    "  --8086              make compressed sys work on any 8086\n"
main.cpp:        if (mfx_optarg && strcmp(mfx_optarg, "8086") == 0)
main.cpp:            opt->cpu = opt->CPU_8086;
main.cpp:        opt->cpu = opt->CPU_8086;
main.cpp:        {"8086", 0x10, N, 561},
main.cpp:        {"8086", 0x10, N, 561},
p_armpe.cpp:    use_thumb_stub |= (opt->cpu == opt->CPU_8086); // FIXME
p_com.cpp:                  filter_id < 4 ? "" : (opt->cpu == opt->CPU_8086 ? "CT16I086" : "CT16I286,CT16SUB0"),
p_com.cpp:                  filter_id < 4 ? "" : (opt->cpu == opt->CPU_8086 ? "CT16I087" : "CT16I287,CT16SUB1"),
p_exe.cpp:                  opt->cpu == opt->CPU_8086 ? "N2BX8601" : "N2B28601",
p_exe.cpp:                  opt->cpu == opt->CPU_8086 ? "N2BX8602" : "N2B28602",
p_exe.cpp:                  opt->cpu == opt->CPU_8086 ? "N2DX8601" : "N2D28601",
p_exe.cpp:                  opt->cpu == opt->CPU_8086 ? "N2DX8602" : "N2D28602",
p_exe.cpp:                  opt->cpu == opt->CPU_8086 ? "N2EX8601" : "N2E28601",
p_exe.cpp:                  opt->cpu == opt->CPU_8086 ? "N2EX8602" : "N2E28602",
p_sys.cpp:              opt->cpu == opt->CPU_8086 ? "SYSI0861" : "SYSI2861",
p_sys.cpp:              opt->cpu == opt->CPU_8086 ? "SYSI0862" : "SYSI2862",
p_vmlinz.cpp:        format = UPX_F_ELKS_8086;
```
Then in the `src/stub/src` directtory, do `grep -sr 086 .` then look in those files.  You will see snippets (`section`s) of the de-compressor that differ for 086 versus 286.  For instance in `arch/i086/nrv2b_d8.S`:
```
section         N2BX8601   ## 8 bytes for 8086
                add     cx, cx
                add     cx, cx
                add     cx, cx
                add     cx, cx
section         N2B28601  ## 3 bytes for 80286
                CPU     286
                shl     cx, 4
                CPU     8086
```
Such code highlights the hardware differences that matter.  Just because the input to UPX has four doubling `add` in a row does NOT mean that 8086 is required.  It could easily mean that an old or lazy compiler generated the code, or was not taught about the `shl reg,immediate` opcode.

If the problem is forgetting to specify `--8086` when invoking UPX, then "hardwire" your copy of UPX by wrapping it in a shell script that supplies `--8086`; or, change the code to initialize `opt->cpu = opt->CPU_8086;`

**Comment by hackerb9** [comment]
Thank you for the explanation.

I had never used UPX before today and found the behaviour surprising. I personally won't have a problem in the future because I added --8086 to my project Makefile. I made this feature request because it would be a nice thing in the future for others who are new to UPX.

That is too bad 8086 code is undetectable. But it sounds like there is very little downside, at least in terms of bytes, to using --8086. 

Have you considered making 8086 the default operating mode? That way 8086 executables would not get munged and people who wanted to save a few extra bytes could use a new flag (--286?).
